HIP/hip-514-missing-hts-precompile-functions.md

```diff
@@ -0,0 +1,244 @@
+| `getTokenKey(address, uint)`                    | xx gas             | 0 gas                 |
+| `getNonFungibleTokenInfo(address, uint32)`      | xx gas             | 0 gas                 |
+| `freezeToken(address, address)`                 | xx gas             | 0 gas                 |
+| `unFreezeToken(address, address)`               | xx gas             | 0 gas                 |
+| `grantTokenKyc(address, address)`               | xx gas             | 0 gas                 |
+| `revokeTokenKyc(address, address)`              | xx gas             | 0 gas                 |
+| `pauseToken(address)`                           | xx gas             | 0 gas                 |
+| `unPauseToken(address)`                         | xx gas             | 0 gas                 |
+| `wipeTokenAccount(address, address, uint32)`    | xx gas             | 0 gas                 |
+| `wipeTokenAccount(address, address, uint64[])`  | xx gas             | xx gas / serial token |
+| `updateTokenExpiryInfo(Expiry)`                 | xx gas             | 0 gas                 |
+| `updateTokenInfo(address, HederaToken)`         | xx gas             | 0 gas                 |
+| `updateTokenKeys(address, uint, KeyValue)`      | xx gas             | xx gas / key          |
+
+## Backwards Compatibility
+
+This HIP builds upon existing Solidity structs and Protobuf messages.
+As such, it is in compliance with previous iterations and should not impact compatibility.
+
+## Security Implications
+
+The HIP provides exposure to existing HTS functionality.
+Fundamentally security is still governed by the ledger security logic in combination with the EVM.
+As such, there should be no security implications in decreased security or increased permissions.
+
+## How to Teach This
+
+- Additonal documentation on [IHederaTokenService.sol](https://github.com/hashgraph/hedera-smart-contracts/blob/main/hts-precompile/IHederaTokenService.sol) and [HederaTokenService.sol](https://github.com/hashgraph/hedera-smart-contracts/blob/main/hts-precompile/HederaTokenService.sol)
+- Client code will be provided with a `TokenManageContract.sol` under [hts-precompile](https://github.com/hashgraph/hedera-smart-contracts/tree/main/hts-precompile)
+- SDK examples should be written to highlight solidity token management function calls.
+- Doc site tutorials utilizing SDKs should be written to highlight solidity token management function calls.
+
+## Reference Implementation
+
+
+
+## Rejected Ideas
+
+The option to create a new precompile solidity file was rejected as the additional functions will benefit from the structure and logic of the current functions
+
+## Open Issues
+
+Regarding [### Solidity Function Signatures][#solidity-function-signatures]
+- Is the `CustomFee[]` optimal on `getTokenCustomFees()` or should a coordinated set of `FixedFee[]`, `FractionalFee[]` and `RoyaltyFee[]` be used instead?
```

Comment:
Have we had a discussion regarding this? What are the advantages from introducing the `CustomFee` wrapper?  

When a caller receives the `CustomFee[]` from a `getTokenCustomFees()`, they'll have to check which of the three fee structs in the `CustomFee` is initialized, which would add some unnecessary calculations, right? Or is that preferred to returning 3 arrays for each type of fee from `getCustomFees()` and also having 3 arrays in the `TokenInfo` struct, in regards to gas optimization?

Maybe @shemnon / @Nana-EC could provide a more detailed insight?

Reply:
I think having 3 separate arrays (`FixedFee[]`, `FractionalFee[]`, `RoyaltyFee[]`) would be more memory saving compared to having a struct (`CustomFee`) that includes 3 separate fields, so I'd also suggest using 3 separate arrays instead of `CustomFee[]`.

@shemnon , @Nana-EC  what do you think?

Reply:
It is preferred in terms of cleanliness to have the struct.
Gas optimization wise given the contract doesn't store the struct my impression was the impact would not be much.

In worst case scenario the 3 arrays need to be checked anyway.
Will look to see if there's a huge overhead but open to thoughts.

Reply:
Solidity is kind of messed up when it comes to C standard ways to do structs.  Particularly in the lack of union types.

The size of the input will indirectly impact gas, it's memory expansion cost, which is 3 gas per null, so requiring two extra nulls per arg will cause 6 gas extra per entry.

From a gas golf perspective the three arrays are cheaper, and more of a standard practice.

Reply:
Updated to 3 arrays and removed `CustomFee` struct